### ISSUE TITLE
BCELoss - weight parameter shape incorrect

### ISSUE THREAD

**Issue by ptrblck** [pytorch/pytorch#5157]
The `weight` parameter of `BCELoss` seems to be incorrectly defined when using a multi-dimensional input and target. Related [forum thread](https://discuss.pytorch.org/t/binary-cross-entropy-weights/13299).

The documentation defines `weight` as:
> If given, has to be a Tensor of size “nbatch”.

However, this example throws an error:
```
x = Variable(torch.randn(4, 2, 2))
y = Variable(torch.Tensor(4, 2, 2).random_(2))
output = F.sigmoid(x)

# Create weight according to doc in BCELoss
weight = torch.randn(4)
criterion_weighted = nn.BCELoss(weight=weight)
loss_weighted = criterion_weighted(output, y) # Error!
```

A workaround is to `unsqueeze` the `weight` tensor to match the number of dimensions:
```
# Unsqueeze weight tensor
weight = torch.randn(4, 1, 1)
criterion_weighted = nn.BCELoss(weight=weight)
loss_weighted = criterion_weighted(output, y)
```

Internally, `_infer_size` is called and fails in the first code snippet.
The second code snippet applied a weighting for each batch element, which is fine.

Should we automatically unsqueeze the weight tensor, if input and target are multi-dimensional?

Also, how should we handle class weighting?
If we just pass 2 weights as the `weight` tensor, the code successfully runs, but does not apply class weighting, which might mislead some users:
```
# Create class weights
weight = torch.FloatTensor([0.1, 0.9]) 

# Internally, weight is expanded as
size = _infer_size(weight.size(), y.size())
weight_expanded = weight.expand(size) 
print(weight_expanded) # This is not, what we wanted as class weights!

criterion_weighted = nn.BCELoss(weight=weight)
loss_weighted = criterion_weighted(output, y)
criterion_nonreduced = nn.BCELoss(reduce=False)
loss_unreduced = criterion_nonreduced(output, y)
loss_weighted_manual = (Variable(weight_expanded) * loss_unreduced).mean()

if loss_weighted == loss_weighted_manual:
    print('Class weighting failed')
```

A workaround would be:
```
weight_ = weight[y.data.view(-1).long()].view_as(y)
criterion = nn.BCELoss(reduce=False)
loss = criterion(output, y)
loss_class_weighted = loss * Variable(weight_)
```

What is wanted behavior of the `weight` parameter in `BCELoss`: class or batch weighting?
Both cases have some issues at the moment in my opinion.
I would like to fix this issue, but I would like to hear some opinions on the right behavior.
Class weighting would be consistent with other loss functions like `NLLLoss`, but maybe batch weighting is a more common use case for `BCELoss`.

PyTorch version: `0.4.0a0+492e26f` (installed from source)

**Comment by zou3519** [comment]
Good catch, @ptrblck!

Based on how BCELoss [is implemented](https://github.com/pytorch/pytorch/blob/ccf4dc15250534e6906aaca696f94b430a4da443/aten/src/THNN/generic/BCECriterion.c#L17), the weight look like they should be per-element weights, not class weights. Correct me if I'm wrong, but I don't think BCELoss has a concept of per-class weights because the inputs and targets are values between 0 and 1.

In this respect the current behavior makes sense (weights of size `(4,)` are not broadcastable to the input size of `(4, 2, 2)`). I think the best thing to do is to fix the docs.

**Comment by ptrblck** [comment]
Thanks for the response @zou3519 !
Yeah, you are right. Class weighting won't work for float targets. I just compared it to other loss functions and was wondering if class weighting would be useful for an imbalanced binary classification problem. `NLLLoss` is probably more useful in this setting. 

Do you think the method should automatically check, if `weight` is for batch or element weighting, i.e. `[4]` or `[4, 2, 2]`, and broadcast it if necessary?

**Comment by zou3519** [comment]
If I'm understanding this correctly, let's say we have inputs and targets of size [4, 2, 2]. You are proposing that if someone sends in a weight of size [4], we expand it to size (4, 2, 2) (or something like that).

What would you say the use case of this is? I think it's more common that people look for per-element weights but if there is a use case this would be a good thing to have.

**Comment by ptrblck** [comment]
I thought about a use case where you would like to weight different instances of the current batch. Something like [sample importance](https://openreview.net/pdf?id=r1IRctqxg). 
However, this could also be achieved using `reduce=False` and a weight matrix.

In the other use case using per-element weights, one would need to reconstruct the `BCELoss` class in each iteration with its new weight matrix? Is it a cheap operation?

What do you think about changing the docs to:
`` 
weight (Tensor, optional) – a manual rescaling weight given to the loss of each element. If given, has to be same shape as the input.
``

**Comment by ptrblck** [comment]
@zou3519 What do you think about multi-label classification using BCELoss?
It seems the `weights` can be used as a class weight in this case.
Related to [this](https://discuss.pytorch.org/t/what-is-the-difference-between-bcewithlogitsloss-and-multilabelsoftmarginloss/14944) thread:

```
x = Variable(torch.randn(10, 3))
y = Variable(torch.FloatTensor(10, 3).random_(2))

# double the loss for class 1
class_weight = torch.FloatTensor([1.0, 2.0, 1.0])
# double the loss for last sample
element_weight = torch.FloatTensor([1.0]*9 + [2.0]).view(-1, 1)
element_weight = element_weight.repeat(1, 3)

bce_criterion = nn.BCEWithLogitsLoss(weight=None, reduce=False)
multi_criterion = nn.MultiLabelSoftMarginLoss(weight=None, reduce=False)

bce_criterion_class = nn.BCEWithLogitsLoss(weight=class_weight, reduce=False)
multi_criterion_class = nn.MultiLabelSoftMarginLoss(weight=class_weight, reduce=False)

bce_criterion_element = nn.BCEWithLogitsLoss(weight=element_weight, reduce=False)
multi_criterion_element = nn.MultiLabelSoftMarginLoss(weight=element_weight, reduce=False)

bce_loss = bce_criterion(x, y)
multi_loss = multi_criterion(x, y)

bce_loss_class = bce_criterion_class(x, y)
multi_loss_class = multi_criterion_class(x, y)

bce_loss_element = bce_criterion_element(x, y)
multi_loss_element = multi_criterion_element(x, y)

print(bce_loss - multi_loss)
print(bce_loss_class - multi_loss_class)
print(bce_loss_element - multi_loss_element)
```

**Comment by sebastian-steindl** [comment]
Any updates on this? @zou3519 @ptrblck 
Issue still remains and I would like to use it to assign weights to every example based on the labelset frequency in my multilabel classification problem. But like @ptrblck mentioned the only way I can get it to work is if I reconstruct the loss object for every batch
